### DISCUSSION TITLE
Define Aliases for `_id`

### DISCUSSION THREAD

**Discussion by koolamusic** [typegoose/typegoose#538] [Q&A] [answered]
Was curious to ask, whether it is possible to define a customer alias for the mongoose `_id` type

pseudo code will be something like 

```ts
export class Customer {
  @prop({ aliases: 'customerId', required: true, default: mongoose.Schema.Types.ObjectId })
  public _id!: mongoose.Types.ObjectId;

  @prop({ required: true, unique: true })
  public userId!: string;

  @prop({ required: true })
  public address!: string;

```

**Comment by hasezoey** [comment, marked as answer]
mongoose provides an way for aliases (virtual, not actually saved as duplicates) with the option `alias`, though it can only work as one alias, not multiple, [there is also documentation on the typegoose side](https://typegoose.github.io/typegoose/docs/api/decorators/prop#alias) on how to best integrate it with typegoose

**Reply by koolamusic** [reply]
thanks. testing this out right now. It seems I cannot create an alias for the `_id` type for some reason

**Reply by koolamusic** [reply]
running `@typegoose/typegoose": "8.0.0-beta.5`

**Reply by koolamusic** [reply]
I am getting a 

```sh

    "message": "customer validation failed: _id: Path `_id` is required."

```

**Reply by hasezoey** [reply]
this is normal, because you are overwriting the default configuration for `_id` - this means that the default id generation also gets "lost"

i dont know if this will help (i just got to know this option recently), but you can try to add `auto: true`, otherwise you will need to define an `default` option yourself (or provide an id on creation of the document)

**Reply by koolamusic** [reply]
ok, let me try that

**Reply by koolamusic** [reply]
setting `{auto: true}` in the `@prop` decorator doesn't work either

**Reply by koolamusic** [reply]
Also I tried creating the `_id` using the mongoose schema type

```ts
CustomerModel.create({
        ...data,
        _id: mongoose.Schema.Types.ObjectId,
      });
 ```

**Reply by koolamusic** [reply]
I am getting this error

```json
{
    "message": "customer validation failed: _id: Cast to ObjectId failed for value \"[Function: ObjectId] {\n   
schemaName: 'ObjectId',\n  defaultOptions: {},\n  get: [Function],\n  set: [Function: set],\n  _checkRequired: 
[Function],\n  _cast: [Function: castObjectId],\n  cast: [Function: cast],\n  _defaultCaster: [Function],\n  
checkRequired: [Function]\n}\" at path \"_id\""
}
```

**Reply by koolamusic** [reply]
False alarm... I think i figured out the issue, I seem to be mixing up the mongoose `Schema` with the mongoose `Types`

**Reply by smolinari** [reply]
Can you post your solution?

Scott

**Reply by koolamusic** [reply]
```ts

import mongoose from 'mongoose';
import { getModelForClass, prop, modelOptions } from '@typegoose/typegoose';

@modelOptions({ options: { customName: 'customer' } })
export class CustomerCollection {
  public customerId!: string;

  @prop({
    required: true,
    alias: 'customerId',
    type: mongoose.Schema.Types.ObjectId,
    default: mongoose.Types.ObjectId(),
  })
  public _id!: mongoose.Schema.Types.ObjectId | string;

  @prop({ required: true })
  public address!: string;

  @prop({ required: true })
  public city!: string;

  @prop({ required: true })
  public state!: string;

  @prop({ required: true })
  public phone!: number;
}

export const CustomerModel = getModelForClass(CustomerCollection);
```


Above is the model definition file, for some reason adding `{auto: true}` doesn't guarantee I will get a default `_id` so I have define `_id` when using customer model.create which isn't an issue, I'm sure a `@pre` hook might be able to take that off my hands as well. See the Create Method example here..


```ts
import { DocumentType, mongoose } from '@typegoose/typegoose';
import { CustomerModel } from './model'

const customer: DocumentType<CustomerCollection> = CustomerModel.create({
        ...sanitizedData,
        _id: mongoose.Types.ObjectId(),
      });
      
```

**Reply by koolamusic** [reply]
I had to define `public customerId!: string` to get types working, although I don't use it in find query parameters at the moment and still stick to using `_id` when I need to make a query
